### PR TITLE
 [Feature] Add KeyStore interface for persistent Proving and Verifying Key storage

### PR DESCRIPTION
**Pull request by iamalwaysuncomfortable**
## Motivation

Aleo functions require a proving key and verifying key per function. On execution, if the keys do not currently exist, they are synthesized and stored in the local `Process` object. However This synthesis process takes a long time and execution time can be cut down drastically by inserting pre-synthesized keys. Normally on executors with filesystems, these keys are stored locally on disk automatically by `SnarkVM`. However `wasm` has no direct access to a filesystem, thus proving and verifying keys must be passed into `wasm` externally by JS. 

Because JS can operate in multiple contexts (such as the `browser`, `node.js` on a machine with a filesystem, or perhaps an enclave environment) key storage will necessarily need to be different depending on the operational context.

Currently there no convenient abstractions in the SDK for managing proving and verifying keys and it is not clear to many users how to do proper proving and verifying key management. This PR adds a `KeyStore` abstraction that provides methods for storing and retrieving proving and verifying keys. A future PR will provide a default implementation called the `FileSystemKeyProvider` for NodeJS sdk users that have access to filesystems. 

This PR also adds the `KeyStore` object to the `FunctionKeyProvider` interface for convenient access.

## Test Plan

* Several unit tests should be written against the `FileSystemKeyProvider` to ensure the `KeyStore` interface operates as desired.
* Unit tests should test the new keyStore() method against the `OfflineKeyProvider` and `AleoKeyProvider` to ensure these interfaces work as desired.

---

> [!NOTE]
> **Medium Risk**
> Touches key-management APIs and adds filesystem persistence/verification logic that can affect how proving/verifying keys are located and validated across environments. Risk is mitigated by comprehensive new unit tests, but path/IO edge cases and browser crypto availability should be watched.
> 
> **Overview**
> **Adds a persistent `KeyStore` abstraction and a Node.js disk-backed implementation.** `FunctionKeyProvider` now exposes `keyStore(): Promise<KeyStore | undefined>`, and a new `LocalFileKeyStore` stores key bytes plus fingerprint metadata under a `.aleo` directory, with atomic writes, path-traversal-safe locator validation (`InvalidLocatorError`/`InvalidLocatorReason`), and optional checksum/size verification on read.
> 
> **Refactors key types/APIs and expands exports/tests.** `FunctionKeyPair`/`CachedKeyPair` move to `models/keyPair.ts`, key-provider interfaces move under `keys/provider/*`, and `browser.ts`/`program-manager.ts`/docs are updated accordingly; `node.ts` exports `LocalFileKeyStore`, and new `KeyVerifier`/`MemKeyVerifier` + `sha256Hex` utilities (re-exported as `ChecksumMismatchError`/`FunctionKeyVerifier`) are added with extensive unit tests covering keystore behavior, restart/metadata verification, corruption detection, and locator validation.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e0109f3bf741dea6dde25c81fce1cbdd513ecaf6. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by iamalwaysuncomfortable** [comment]
@cursor review

File: sdk/src/keys/keystore/file.ts [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+
+    /**
+     * Generates the path for a key metadata file based on the locator.
+     *
+     * @private
+     * @param {string} locator - Unique identifier for the key.
+     * @returns {string} Full filesystem path to the metadata file.
+     */
+    private metadataPath(locator: string): string {
+        return path.join(this.directory, `${locator}.metadata`);
```
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply sdk/src/keys/keystore/file.ts:76]: Resolved by validating locators at their entry points.

File: sdk/src/keys/keystore/file.ts [hunk 2]
```diff
@@ -0,0 +1,420 @@
+import * as fs from "node:fs/promises";
+import * as path from "path";
+
+import { FunctionKeyPair } from "../../models/keyPair.js";
+import { KeyFingerprint, } from "../verifier/interface.js";
+import { KeyLocator, KeyStore } from "./interface.js";
+import { MemKeyVerifier } from "../verifier/memory.js";
+import { ProvingKey, VerifyingKey } from "../../wasm.js";
+
+export class LocalFileKeyStore implements KeyStore {
```
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply sdk/src/keys/keystore/file.ts:12]: PR description updated, this will be done in a follow-on PR.

File: sdk/src/keys/keystore/file.ts [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+     *   locator: 'transfer_private.prover.421e5a5',
+     *   fingerprint: { checksum: '421e5a52f01a1eeb068bbf56d15e19477ff7290e4b988d1013e15563f2b77801', size: 116746954 }
+     * });
+     * if (keyBytes) {
+     *   // Use the verified key bytes
+     * }
+     */
+    async getKeyBytes(locator: KeyLocator): Promise<Uint8Array | null> {
+        // Attempt to read key bytes from storage (under this.directory).
+        const keyBytes = await this.readFileOptional(path.join(this.directory, locator.locator));
```
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply sdk/src/keys/keystore/file.ts:224]: Resolved by validating locators in each method.

File: sdk/src/keys/keystore/file.ts [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+     * @throws {Error} If directory creation fails.
+     */
+    constructor(directory?: string) {
+        this.directory = directory ?? path.join(process.cwd(), "keystore");
+
+        // Ensure directory exists
+        fs.mkdir(this.directory, { recursive: true }).catch((err) => {
+            console.error("Failed to create keystore directory:", err);
+        });
+    }
```
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply sdk/src/keys/keystore/file.ts:66]: The constructor now throws if the directory creation fails (but does NOT throw if it already exists).

File: sdk/src/keys/keystore/file.ts [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+     * @returns {Promise<void>} Resolves when write is complete
+     * @throws {Error} If directory creation or file write fails
+     */
+    private async writeFileAtomic(
+        filepath: string,
+        data: Uint8Array,
+    ): Promise<void> {
+        // Ensure parent directories for nested locators exist
+        await fs.mkdir(path.dirname(filepath), { recursive: true });
+        await fs.writeFile(filepath, data);
```
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply sdk/src/keys/keystore/file.ts]: This has been addressed

File: sdk/src/keys/keystore/file.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply sdk/src/keys/keystore/file.ts:66]: The constructor now fails if the directory can't be created and appends a .aleo to it.

File: sdk/src/keys/keystore/file.ts [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+
+                if (isDirectory) {
+                    await this.clearRecursive(full);
+                } else if (
+                    name.endsWith(".prover") ||
+                    name.endsWith(".verifier") ||
+                    name.endsWith(".metadata")
+                ) {
+                    await fs.unlink(full).catch(() => {});
+                }
```
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply sdk/src/keys/keystore/file.ts]: This has been addressed by enforcing all files MUST be in the .aleo directory

File: sdk/src/keys/keystore/file.ts [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+        // If a fingerprint was provided, verify the key bytes against it.
+        if (
+            locator.fingerprint ||
+            (await this.getKeyMetadata(locator.locator))
+        ) {
+            await this.keyVerifier.verifyKeyBytes({
+                keyBytes,
+                locator: locator.locator,
+                fingerprint: locator.fingerprint,
+            });
```
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply sdk/src/keys/keystore/file.ts:237]: Excellent catch!

File: sdk/src/keys/keystore/file.ts [hunk 8]
```diff
@@ -0,0 +1,477 @@
...
+            }
+            throw err;
+        }
+
+        await Promise.all(
+            entries.map(async (name) => {
+                const full = path.join(dir, name);
+                let stat: Awaited<ReturnType<typeof fs.stat>>;
+                try {
+                    stat = await fs.stat(full);
```
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply sdk/src/keys/keystore/file.ts]: This has been fixed.
